// message-controller/HandleMessage.js
/*
	Copyright 2018 [4ddev *see github.com/4ddev]

	Permission is hereby granted, free of charge, to any person obtaining a copy 
	of this software and associated documentation files (the "Software"), 
	to deal in the Software without restriction, including without limitation 
	the rights to use, copy, modify, merge, publish, distribute, sublicense, 
	and/or sell copies of the Software, and to permit persons to whom the Software 
	is furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be 
	included in all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
	EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
	OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
	DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, 
	ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR 
	OTHER DEALINGS IN THE SOFTWARE.
*/
class Set{
	constructor(){
		this.keys = [];
	}
	// Stores Classes
	put(){
		for( let i = 0; i < arguments.length; i++){
		   if( arguments[i].name != null ) this.keys[arguments[i].name] = arguments[i];
		}
		return this;
	}
	contains( obj ){
		return this.keys[obj]!=null;
	}
	
	remove( obj ){
		this.keys[obj] = null;
		delete this.keys[obj];
	}
	get( key ){
		return this.keys[key];
	}
}

class HandleMessage{
	/* Does not attach function's or their return values just atomic values */
	static wrapMessage( obj ){
		return "{\""+obj.constructor.name+"\":"+JSON.stringify( obj  )+"}"; 
	}
	/*
	   @param json Some Object or an String which will be parsed to an Object by json.parse 
	   @param classes Check's incoming Message against this ClassSet if no class with the name was 
	                  found, it will remove null 
       @return null If the class was not found
	           Object The instantianted Object in the Classlist 
	*/			   
	static createInstance(json,classes){
		if ( json.constructor != null && json.constructor.name == "String" ){ 
			try{
				json = JSON.parse(json); 
			}catch( e ){
				throw new Error("Could not parse JSON"+json);
			}

		}
		let keys = Object.keys(json);
		if( keys.length != 1 ) throw new Error ("Manipulated Command!");
		if (classes.contains(keys[0]) ) { 
			return Object.assign(new (classes.get(keys[0]))(),json[keys[0]]); 
		}
		return null;
		//throw new Error("Method is not allowed"+ Object.keys(json)[0]);  // Use this if you wanna have an exception 
	}
}

module.exports = {
	HandleMessage: HandleMessage,
	ClassSet: Set,
	createInstance: HandleMessage.createInstance,
	wrapMessage: HandleMessage.wrapMessage,
}
